src/base/utils/version.cpp

```diff
@@ -0,0 +1,33 @@
+/*
+ * Bittorrent Client using Qt and libtorrent.
+ * Copyright (C) 2023  Mike Tzou (Chocobo1)
+ *
+ * This program is free software; you can redistribute it and/or
+ * modify it under the terms of the GNU General Public License
+ * as published by the Free Software Foundation; either version 2
+ * of the License, or (at your option) any later version.
+ *
+ * This program is distributed in the hope that it will be useful,
+ * but WITHOUT ANY WARRANTY; without even the implied warranty of
+ * MERCHANTABILITY or FITNESS FOR A PARTICULAR PURPOSE.  See the
+ * GNU General Public License for more details.
+ *
+ * You should have received a copy of the GNU General Public License
+ * along with this program; if not, write to the Free Software
+ * Foundation, Inc., 51 Franklin Street, Fifth Floor, Boston, MA  02110-1301, USA.
+ *
+ * In addition, as a special exception, the copyright holders give permission to
+ * link this program with the OpenSSL project's "OpenSSL" library (or with
+ * modified versions of it that use the same license as the "OpenSSL" library),
+ * and distribute the linked executables. You must obey the GNU General Public
+ * License in all respects for all of the code used other than "OpenSSL".  If you
+ * modify file(s), you may extend this exception to your version of the file(s),
+ * but you are not obligated to do so. If you do not wish to do so, delete this
+ * exception statement from your version.
+ */
+
+#include "version.h"
+
+#include "base/interfaces/istringable.h"
+
+static_assert(Stringable<Utils::Version<1>>);
```

Comment:
>Another instantiations still may not satisfy Stringable...

As said before, this is unsolvable.

>BTW, is static_assert unacceptable to be in header file?

Then every direct/indirect include will evaluate once which will unnecessary slow down compilation.

Reply:
Now I have another idea. Maybe, after all, not to pollute the source code with such "unnatural" assertions, but simply add a test that will check all such classes?

Reply:
>Now I have another idea. Maybe, after all, not to pollute the source code with such "unnatural" assertions, but simply add a test that will check all such classes?

I thought of it before but I consider it to be as unnatural as such assertions, i.e. it isn't better.

Reply:
> > Another instantiations still may not satisfy Stringable...
> 
> As said before, this is unsolvable.

Can't you put `static_assert(Stringable<Version<N>>);` within class (template) declaration (near other `static_assert`'s)?

Reply:
> > Now I have another idea. Maybe, after all, not to pollute the source code with such "unnatural" assertions, but simply add a test that will check all such classes?
> 
> I thought of it before but I consider it to be as unnatural as such assertions, i.e. it isn't better.

Okay, let's not continue this argument, let's stay (as a compromise) with the option with "exculpatory" comments near the assertions.

Reply:
>Can't you put static_assert(Stringable<Version<N>>); within class (template) declaration (near other static_assert's)?

Please see it yourself https://godbolt.org/z/vfzPcrhT5

Reply:
> I've said it numerous times

I don't recall you mentioning that you checked this option too...

But now I noticed something else... `Version` didn't satisfy `Stringable` before (it had no `Version(QString)` compatible constructor), which means it wasn't required. So why did you add this requirement?

Reply:
>I don't recall you mentioning that you checked this option too...

I think I implied it in my previous replies that those forms are not valid.

>But now I noticed something else... Version didn't satisfy Stringable before (it had no Version(QString) compatible constructor), which means it wasn't required. So why did you add this requirement?

IIRC this is intended in PR #16738 and however was neglected during redesign in #17484 and now I'm fixing it.

Reply:
I meant that we don't store `Version` in the `SettingsStorage` (otherwise our current code wouldn't compile). So why add such a requirement? As mentioned above, if a class satisfies a concept, it is not obliged to declare it. We agreed that we are doing this as a workaround in the specific case where some custom type is intended to be stored in `SettingsStorage`. Or is it still intended despite the fact that it hasn't been used yet?